lib/registration.cpp

```diff
@@ -13,4 +20,309 @@ transform_pointcloud_inplace(EigenPointCloudRef cloud,
+supervoxel_segmentation(Epoch& epoch,
+                        const KDTree& kdtree, //?
+                        double resolution,
+                        int k)
+{
+
+  // Define number of supervoxels and labels.
+  auto n_supervoxels = estimate_supervoxel_count(epoch.cloud, resolution);
+  std::vector<int> labels(epoch.cloud.rows(), -1);
+  DisjointSet set(epoch.cloud.rows());
+
+  // Calculate normals for vccs metric
+  epoch.kdtree.build_tree(10);
+  std::vector<double> normal_radii{ 2.0 };
+  EigenNormalSet normals(epoch.cloud.rows(), 3);
+  EigenNormalSet orientation(1, 3);
+  orientation << 0, 0, 1;
+  compute_multiscale_directions(
+    epoch, epoch.cloud, normal_radii, orientation, normals);
+
+  normals.norm();
+
+  // Calculate k neigbors and its distances for each point
+  KDTree::NearestNeighborsDistanceResult result;
+  kdtree.nearest_neighbors_with_distances(epoch.cloud, result, k);
+
+  int supervoxels_amount = epoch.cloud.rows();
+
+  // calculate lambda for segmentation
+  DistanceVector lambda_distances;
+  for (size_t i = 0; i < result.size(); ++i) {
+    std::nth_element(result[i].second.begin(),
+                     result[i].second.begin() + 1,
+                     result[i].second.end());
+    lambda_distances.push_back(result[i].second[1]);
+    // because [0] is the distance to the same point, so it's 0
+  }
+
+  double lambda = median_calculation(lambda_distances);
+
+  // initialize temporary vars for supervoxel segmentation
+  std::vector<int> temporary_supervoxels(epoch.cloud.rows());
+  std::iota(temporary_supervoxels.begin(), temporary_supervoxels.end(), 0);
+  std::vector<int> sizes(epoch.cloud.rows(), 1); //+-?
+  std::queue<int> point_queue;
+  std::vector<std::vector<long unsigned int>> neighborIndexes(result.size());
+
+  // fill in temporal containment of neighbors for each points
+  for (size_t i = 0; i < result.size(); ++i) {
+    for (const long unsigned int& index : result[i].first) {
+      neighborIndexes[i].push_back(index);
+    }
+  }
+
+  DistanceVector distances;
+  // searching for supervoxels and first segmentation
+  for (; lambda < LMBD_MAX; lambda *= 2.0) {
+    for (auto i : temporary_supervoxels) {
+      if (neighborIndexes[i].empty())
+        continue;
+
+      labels[i] = i;
+      point_queue.push(i);
+
+      for (auto j : neighborIndexes[i]) {
+        j = set.Find(j);
+        if (labels[j] == -1) {
+          point_queue.push(j);
+        }
+      }
+
+      std::vector<long unsigned int> neighborIndexes_per_point;
+      while (!point_queue.empty()) {
+        int current = point_queue.front();
+        point_queue.pop();
+
+        double loss =
+          sizes[current] *
+          squaredEuclideanDistance(
+            epoch.cloud.row(i),
+            epoch.cloud.row(
+              current)); // point_2_point_VCCS_distance(epoch.cloud.row(i),epoch.cloud.row(current),
+                         // normals.row(i), normals.row(current), resolution);
```

Comment:
What would be required to switch the metric here. I just want to find out to communicate with Yihui and the others about our requirements.